[WIP] Revamped octane tutorial (#1002)
ember-learn/guides-source: guides/release/tutorial/8.md

We accomplished a lot in the previous chapter! Let's recap &mdash; we started with a blank canvas, added a few pages of content, styled everything to look pretty, dropped in a picture of Tomster, added links between our pages and amazingly, everything worked together flawlessly!

But do we _really_ know that everything is actually working? Sure, we clicked around a bit to confirm that things look as expected. But do we feel confident that we checked _every_ page after the most recent change that we made?

> After all, most of us have experienced (or heard horror stories about) making a Small Tweak™ in one area of the app and inadvertently breaking _everything else_ when we weren't looking.

Reviewer: ```suggestion
After all, most of us have experienced (or heard horror stories about) making a Small Tweak™ in one area of the app that inadvertently breaks _everything else_ when we aren't looking.
```

> Maybe we can write a checklist somewhere of all the things to check after making changes to our site. But surely, this will get out of hand quickly as we add more features to our app? It is also going to get old really quickly &mdash; repetitive tasks like that are best left to robots.

Reviewer: ```suggestion
Maybe we can write a checklist somewhere of all the things to check after making changes to our site. But surely, this will get out of hand as we add more features to our app. It is also going to get old really quickly &mdash; repetitive tasks like that are best left to robots.
```

In this case, we generated an *[acceptance test][TODO: link to acceptance test]* located at `tests/acceptance/super-rentals.js`.

Generators aren't required; we _could_ have created the file ourselves which would have accomplished the exact same thing. But, generators certainly save us a lot of typing. Go ahead and take a peek at the acceptance test file and see for yourself.

<div class="cta">
  <div class="cta-note">
    <div class="cta-note-body">
      <div class="cta-note-heading">Zoey says...</div>
      <div class="cta-note-message">
        Want to save even more typing? `ember generate ...` can be shortened into `ember g ...`. That's 7 fewer characters!
      </div>
    </div>
    <img src="/images/mascots/zoey.png" role="presentation" alt="Ember Mascot">
  </div>
</div>

> Acceptance tests are one of the few types of automated testing at our disposal in Ember. We will learn about the other types later, but what makes acceptance tests unique is that they test our app from the user's perspective &mdash; they are an automated version of the "click around and see if it works" testing we did earlier, which is exactly what we need.

Reviewer: ```suggestion
Acceptance tests are one of a few types of automated testing at our disposal in Ember. We will learn about the other types later, but what makes acceptance tests unique is that they test our app from the user's perspective &mdash; they are an automated version of the "click around and see if it works" testing we did earlier, which is exactly what we need.
```

Let's open the generated test file and replace the boilerplate test with our own:

<-- TODO: format diff -->
```
import { module, test } from 'qunit';
import { click, visit, currentURL } from '@ember/test-helpers';
import { setupApplicationTest } from 'ember-qunit';

 module('Acceptance | super rentals', function(hooks) {
  setupApplicationTest(hooks);

   test('visiting /', async function(assert) {
    await visit('/');

    assert.equal(currentURL(), '/');
    assert.dom('h2').containsText('Welcome to Super Rentals!');

    assert.dom('.jumbo a.button').containsText('About Us');
    await click('.jumbo a.button');

    assert.equal(currentURL(), '/about');
  });
});
```

First, we instruct the test robot to navigate to the `/` URL of our app by using the `visit` _test helper_ provided by Ember. This is akin to us typing `http://localhost:4200/` in the browser's address bar and hitting the `enter` key.

> Because the page is going to take some time to load, this is known as an *[async](TODO: link to async)* (short for _asynchronous_) step, so we will need to tell the test robot to wait by using the `await` keyword in JavaScript. That way, it will know to wait until the page finishes loading completely before moving on to the next step.

Reviewer: ```suggestion
Because the page is going to take some time to load, this is known as an *[async](TODO: link to async)* (short for _asynchronous_) step, so we will need to tell the test robot to wait by using JavaScript's `await` keyword. That way, it will wait until the page completely finishes loading before moving on to the next step.
```

> This is almost always the behavior you want, so you will almost always use `await` and `visit` as a pair. This applies to other kinds of simulated interaction too, such as clicking on a button or a link, as they all takes time to complete. Even though sometimes these actions may seem imperceptibly fast to us, we have to remember that our test robot has really really fast hands, as we will see in a moment.

Reviewer: ```suggestion
This is almost always the behavior we want, so we will almost always use `await` and `visit` as a pair. This applies to other kinds of simulated interaction too, such as clicking on a button or a link, as they all take time to complete. Even though sometimes these actions may seem imperceptibly fast to us, we have to remember that our test robot has really really fast hands, as we will see in a moment.
```

> After navigating to the `/` URL and waiting for things to settle, we check that the current URL matches the URL that we expect (`/`). We can use the `currentURL` test helper here, as well as the `equal` *[assertion](TODO: link to assertion)*. This is how we encode our "checklist" into code &mdash; by specifying, or *[asserting](TODO: link to asserting)* how things should behave, we will be notified when our app does not behave in the way that we expect.

Reviewer: ```suggestion
After navigating to the `/` URL and waiting for things to settle, we check that the current URL matches the URL that we expect (`/`). We can use the `currentURL` test helper here, as well as `equal` *[assertion](TODO: link to assertion)*. This is how we encode our "checklist" into code &mdash; by specifying, or *[asserting](TODO: link to asserting)* how things _should_ behave, we will be alerted if our app does _not_ behave in the way that we expect.
```

> Next, we confirmed that there should be a `<h1>` tag on the page that contains the text "Welcome to Super Rentals!". Knowing this is true means that we can be quite certain that the correct template is being rendered, without errors.

Reviewer: ```suggestion
Next, we confirmed that the page has an `<h1>` tag that contains the text "Welcome to Super Rentals!". Knowing this is true means that we can be quite certain that the correct template has been rendered, without errors.
```

> Then, we looked for a link with the text `About Us`, located using the *[CSS selector](TODO: link to CSS selector)* `.jumbo a.button`. This is the same syntax we used in our stylesheet, which means "look inside the tag with the `jumbo` class, for an `<a>` tag with the `button` class". This matches up with the HTML structure in our template.

Reviewer: ```suggestion
Then, we looked for a link with the text `About Us`, located using the *[CSS selector](TODO: link to CSS selector)* `.jumbo a.button`. This is the same syntax we used in our stylesheet, which means "look inside the tag with the `jumbo` class for an `<a>` tag with the `button` class". This matches up with the HTML structure in our template.
```

> Once the existence of this element on the page is confirmed, we told the test robot to click on this link. As mentioned above, this is a user interaction, so it needs to be `await`-ed.

Reviewer: ```suggestion
Once the existence of this element on the page was confirmed, we told the test robot to click on this link. As mentioned above, this is a user interaction, so it needs to be `await`-ed.
```

Finally, we asserted that clicking on the link should bring us to the `/about` URL.

<div class="cta">
  <div class="cta-note">
    <div class="cta-note-body">
      <div class="cta-note-heading">Zoey says...</div>
      <div class="cta-note-message">

>         Here, we are writing the tests in a framework called QUnit, which is where `module`, `test` and `assert` comes from. We also had additional helpers like `click`, `visit` and `currentURL` provided by the `@ember/test-helpers` package. You tell what came from which package based on the `import` paths at the top of the file. Knowing this will be helpful when you need to search for documentation on the Internet or ask for help.

Reviewer: ```suggestion
        Here, we are writing the tests in a framework called QUnit, which is where the functions `module`, `test` and `assert` come from. We also have additional helpers like `click`, `visit`, and `currentURL` provided by the `@ember/test-helpers` package. You can tell what comes from which package based on the `import` paths at the top of the file. Knowing this will be helpful when you need to search for documentation on the Internet or ask for help.
```

      </div>
    </div>
    <img src="/images/mascots/zoey.png" role="presentation" alt="Ember Mascot">
  </div>
</div>

> We can put our automated test into motion by running the *[test server](TODO: link to test server)*, using the `ember test --server` command, or `ember t -s` for short. This server behaves like the development server, but it is explicitly running for our tests. It may automatically open a browser window and take you to the test UI, or you can open `http://localhost:7357/` yourself.

Reviewer: ```suggestion
We can put our automated test into motion by running the *[test server](TODO: link to test server)* using the `ember test --server` command, or `ember t -s` for short. This server behaves much like the development server, but it is explicitly running for our tests. It may automatically open a browser window and take you to the test UI, or you can open `http://localhost:7357/` yourself.
```

    assert.dom('h2').containsText('Welcome to Super Rentals!');

    assert.dom('.jumbo a.button').containsText('About Us');
    await click('.jumbo a.button');

    assert.equal(currentURL(), '/about');
  });

   test('visiting /about', async function(assert) {
    await visit('/about');

    assert.equal(currentURL(), '/about');
    assert.dom('h2').containsText('About Super Rentals');

    assert.dom('.jumbo a.button').containsText('Contact Us');
    await click('.jumbo a.button');

    assert.equal(currentURL(), '/getting-in-touch');
  });

   test('visiting /getting-in-touch', async function(assert) {
    await visit('/getting-in-touch');

    assert.equal(currentURL(), '/getting-in-touch');
    assert.dom('h2').containsText('Contact Us');

    assert.dom('a.button').containsText('About');
    await click('.jumbo a.button');

    assert.equal(currentURL(), '/about');
  });
});
```

As with the development server, the test UI should automatically reload and rerun the entire test suite as you save the files. It is recommended that you keep this page open as you develop your app. That way, you will get immediate feedback if you accidentally break something.

> For the rest of the tutorial, we will continue to add more automated tests as we develop new features. Testing is optional; tests don't affect the functionality your app, they just protect it from *[regressions](TODO: link to regressions)*, which is just a fancy way of saying accidental breakages.

Reviewer: ```suggestion
For the rest of the tutorial, we will continue to add more automated tests as we develop new features. Testing is optional; tests don't affect the functionality your app, they just protect it from *[regressions](TODO: link to regressions)*, which is just a fancy way of saying "accidental breakages."
```
